File: app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmBroadcastReceiver.java
package de.tutao.tutanota.alarms;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;

import java.util.Date;

import de.tutao.tutanota.MainActivity;
import de.tutao.tutanota.push.LocalNotificationsFacade;

public class AlarmBroadcastReceiver extends BroadcastReceiver {
	public static final String ALARM_NOTIFICATION_CHANNEL_ID = "alarms";
	private static final String TAG = "AlarmBroadcastReceiver";

	private static final String SUMMARY_EXTRA = "summary";
	public static final String EVENT_DATE_EXTRA = "eventDate";

	public static Intent makeAlarmIntent(Context context, int occurrence, String identifier, String summary, Date eventDate, String userId) {
		String occurrenceIdentifier = identifier + "#" + occurrence;
		Intent intent = new Intent(context, AlarmBroadcastReceiver.class);
		intent.setData(Uri.fromParts("alarm", occurrenceIdentifier, ""));
		intent.putExtra(SUMMARY_EXTRA, summary);
		intent.putExtra(EVENT_DATE_EXTRA, eventDate.getTime());
		intent.putExtra(MainActivity.OPEN_USER_MAILBOX_USERID_KEY, userId);
		return intent;
	}

	@Override
	public void onReceive(Context context, Intent intent) {
		Log.d(TAG, "Received alarm broadcast");
		long when = intent.getLongExtra(EVENT_DATE_EXTRA, System.currentTimeMillis());
		String summary = intent.getStringExtra(SUMMARY_EXTRA);
		LocalNotificationsFacade.showAlarmNotification(context, when, summary, intent);
	}
}

File: app-android/calendar/src/main/java/de/tutao/calendar/alarms/AlarmBroadcastReceiver.kt
package de.tutao.calendar.alarms

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import de.tutao.calendar.MainActivity
import de.tutao.calendar.push.showAlarmNotification
import java.util.Date

class AlarmBroadcastReceiver : BroadcastReceiver() {
	override fun onReceive(context: Context, intent: Intent) {
		Log.d(TAG, "Received alarm broadcast")
		val timestamp = intent.getLongExtra(EVENT_DATE_EXTRA, System.currentTimeMillis())
		val summary = intent.getStringExtra(SUMMARY_EXTRA)
		val isAllDayEvent = intent.getBooleanExtra(IS_ALL_DAY_EVENT, false)
		showAlarmNotification(context, timestamp, summary!!, isAllDayEvent, intent)
	}

	companion object {
		private const val TAG = "AlarmBroadcastReceiver"
		private const val SUMMARY_EXTRA = "summary"
		private const val EVENT_DATE_EXTRA = "eventDate"
		private const val IS_ALL_DAY_EVENT = "isAllDayEvent"

		fun makeAlarmIntent(
			context: Context,
			occurrence: Int,
			identifier: String,
			summary: String?,
			eventDate: Date,
			isAllDayEvent: Boolean,
			userId: String?
		): Intent {
			val occurrenceIdentifier = "$identifier#$occurrence"
			val intent = Intent(context, AlarmBroadcastReceiver::class.java)
			intent.data = Uri.fromParts("alarm", occurrenceIdentifier, "")
			intent.putExtra(SUMMARY_EXTRA, summary)
			intent.putExtra(EVENT_DATE_EXTRA, eventDate.time)
			intent.putExtra(MainActivity.OPEN_USER_MAILBOX_USERID_KEY, userId)
			intent.putExtra(IS_ALL_DAY_EVENT, isAllDayEvent)
			return intent
		}
	}
}